lib/session_encryptor.rb

```diff
@@ -1,18 +1,128 @@
 class SessionEncryptor
+  class SensitiveKeyError < StandardError; end
+  SENSITIVE_KEYS = [
+    'first_name', 'middle_name', 'last_name', 'address1', 'address2', 'city', 'state', 'zipcode',
+    'zip_code', 'dob', 'phone', 'phone_number', 'ssn', 'prev_address1', 'prev_address2',
+    'prev_city', 'prev_state', 'prev_zipcode', 'pii', 'pii_from_doc', 'password', 'personal_key'
+  ]
+
+  SENSITIVE_PATHS = [
+    ['warden.user.user.session', 'idv/doc_auth'],
+    ['warden.user.user.session', 'idv'],
+    ['warden.user.user.session', 'personal_key'],
```

Comment:
Did we consider trying to store a completely separate key in redis for the sensitive data, ex outside of Warden/Devise?

then the entire warden/devise session could be unencrypted, and we know not to put PII in there anymore, so then in controllers, etc we'd access `encrypted_session` and write PII there instead? The key for the `encrypted_session` could live inside the unencrypted session?

Reply:
It's definitely a place I'd like to get, but it was hard to untangle `idv/doc_auth` and `idv` in a safe way that was backwards compatible.  The "safe path" configuration kind of does what you're suggesting (in a roundabout way) by deleting all of the sensitive paths and creating a new hash within the session and then later decrypting and re-merging it.

The FSM v2 work plays into this a little bit in that we'll hopefully not be storing anything during IDV. I think we could ideally not be storing personal_key in the session at all too (e.g. render it in the same request that we generate it and store it nowhere).

Another avenue I explored was to have a more structured object (that is more than just a Hash) so we can be a little more precise about how/what we store, but Rails was not cooperating with that.

Reply:
I’ll add that the idv keys are pretty broad and they do contain a good amount of data that isn’t sensitive.

Future versions could to pare it down to be more specific, but this an 80% solution and it may not be worth it with FSMv2 happening.

Reply:
> The FSM v2 work plays into this a little bit in that we'll hopefully not be storing anything during IDV. I think we could ideally not be storing personal_key in the session at all too (e.g. render it in the same request that we generate it and store it nowhere).

Yes, getting personal key out of the session is absolutely necessary, and once we have FSMv2 focusing on only having one place where we need encrypted data in the session seems like the ideal state. I do like that this gets KMS out of the loop for non-IdV requests immediately though.

Reply:
The current KMS call/time which is pure latency is less than the auth/digest time which is CPU bound.  So even though we'd cut down on network calls there would appear to be no scaling benefit

Reply:
the kms encrypt/decrypt still makes a decent dent in cpu so I'm not against this. it's very cool